Library/Homebrew/livecheck/livecheck.rb

```diff
@@ -0,0 +1,384 @@
+# frozen_string_literal: true
+
+module Homebrew
+  module Livecheck
+    module_function
+
+    GITHUB_SPECIAL_CASES = %w[
+      api.github.com
+      /latest
+      mednafen
+      camlp5
+      kotlin
+      osrm-backend
+      prometheus
+      pyenv-virtualenv
+      sysdig
+      shairport-sync
+      yuicompressor
+    ].freeze
+
+    UNSTABLE_VERSION_KEYWORDS = %w[
+      alpha
+      beta
+      bpo
+      dev
+      experimental
+      prerelease
+      preview
+      rc
+    ].freeze
+
+    def livecheck_formulae(formulae_to_check, args)
+      # Identify any non-homebrew/core taps in use for current formulae
+      non_core_taps = {}
+      formulae_to_check.each do |f|
+        next if f.tap.blank?
+        next if f.tap.name == "homebrew/core"
+
+        non_core_taps[f.tap.name] = true
+      end
+      non_core_taps = non_core_taps.keys.sort
+
+      # Load additional Strategy files from taps
+      non_core_taps.each do |tap_name|
+        tap_strategy_path = "#{Tap.fetch(tap_name).path}/livecheck/strategy"
+        Dir["#{tap_strategy_path}/*.rb"].sort.each(&method(:require)) if Dir.exist?(tap_strategy_path)
+      end
+
+      # Cache demodulized strategy names, to avoid repeating this work
+      @livecheck_strategy_names = {}
+      Strategy.constants.sort.each do |strategy_symbol|
+        strategy = Strategy.const_get(strategy_symbol)
+        @livecheck_strategy_names[strategy] = strategy.name.demodulize
+      end
+      @livecheck_strategy_names.freeze
+
+      has_a_newer_upstream_version = false
+      formulae_checked = formulae_to_check.sort.map.with_index do |formula, i|
+        puts "\n----------\n" if args.debug? && i.positive?
+
+        skip_result = skip_conditions(formula, args: args)
+        next skip_result if skip_result != false
+
+        formula.head.downloader.shutup! unless formula.stable?
+
+        current = formula.stable? ? formula.version : formula.installed_version.version.commit
```

Comment:
What does `formula.version` return for HEAD-only formulae?

Reply:
I created a temporary HEAD-only formula to test this, and when not installed:
```ruby
irb(main):001:0> :headonly.f
=> #<Formula headonly (head) /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/headonly.rb>
irb(main):002:0> :headonly.f.version
=> #<HeadVersion:0x00007fab40023300 @version="HEAD", @commit=nil>
```

When installed:
```ruby
irb(main):002:0> :headonly.f.version
=> #<HeadVersion:0x00007facf824d1d0 @version="HEAD", @commit=nil>
irb(main):003:0> :headonly.f.installed_version
=> #<PkgVersion:0x00007facf5c8f600 @version=#<HeadVersion:0x00007facf5c8f6c8 @version="HEAD-92a144d", @commit="92a144d">, @revision=0>
```

Using `formula.installed_version.version.commit` ensures we get the commit ID of the currently installed version.

Reply:
Cool, seems good!